packages/back-end/src/events/base-types.ts

```diff
@@ -0,0 +1,39 @@
+export const APP_NOTIFICATION_EVENT_EMITTER_NAME = "notification_event";
+
+/**
+ * Supported events for event notifications
+ */
+export const notificationEventNames = [
+  "feature.created",
+  "feature.updated",
+  "feature.deleted",
+] as const;
+
+export type NotificationEventName = typeof notificationEventNames[number];
+
+/**
+ * Supported resources for event notifications
+ */
+export const notificationEventResources = ["feature", "experiment"] as const;
+export type NotificationEventResource = typeof notificationEventResources[number];
+
+/**
+ * Event Notification payload
+ */
+export type NotificationEventPayload<
+  EventName extends NotificationEventName,
+  ResourceType extends NotificationEventResource | unknown,
+  DataType
+> = {
+  event_id: string;
+  event: EventName;
+  object: ResourceType;
+  data: DataType;
```

Comment:
For update events, most handlers we add are going to need to know what changed.  For example, in a Slack handler, we will likely construct human-readable strings like this:

> Feature `xyz` was toggled `ON` in production

There's no easy way to get the original value to do a diff since the database has already been updated by the time the handler is called.

This is also relevant for delete events. For example, if a feature is deleted, you might need to know which environments it was previously enabled in so you can rerefresh your SDK endpoint cache.

Reply:
I will address the updated and deleted concerns separately below.

### Updated

I'm wondering if that's something we can handle in the Slack handler, and any relevant handlers that may care about information in that format. For example, it looks like we already store revisions to features in `featurerevisions`, and it looks like there's enough information in that collection to help us calculate that information.

For example, if `feature.updated` event was fired for a feature, we can see it has a `feature.revision.version` integer on it.

<img width="352" alt="image" src="https://user-images.githubusercontent.com/113377031/203407015-7f137157-1e79-4531-8303-0d6d366c41e5.png">

In the Slack handler, we can query this revision and the previous one. We can do so, for example, with the following find query on `featurerevisions`:

```js
{
  featureId: "integer_feature",
  $and: [
    { version: { $lte: 8 } },
    { version: { $gte: 7 } }
  ]
}
```

Dynamically 8 would be `feature.revision.version` and the 7 would be replaced with `feature.revision.version - 1`. 

We could see the latest feature revisions and compare all the properties and see that the `defaultValue` in this case has changed. We would want to probably manually curate this rather than use an automated diff checker because some fields may not be relevant or may want to be handled differently, e.g.:

> The default value for feature **integer_feature** has been changed from 1001 to 999. 
> Comment: I made an oopsie

<img width="365" alt="image" src="https://user-images.githubusercontent.com/113377031/203408391-f60a368d-f528-4270-b48b-9ffad80b78c8.png">

Since this kind of messaging would be relevant to Slack, and may not be relevant to others, what if we left it as is?

Alternatively, maybe we can fire multiple events:

- the current one I have, for those who just care about the new data
- a new event, e.g. `feature.updated.diff` that has properties `before` and `after`

For Slack, perhaps we would only care to handle the `feature.updated.diff` rather than the `feature.updated` because we may not care about the whole object and only what's changed.

This also means when we want to track other updated events for other resources, we would need to create similar collections to track revisions. If it's important for us to be able to report a diff for collections, then I think storing all versions makes sense.

### Deleted

If we want to examine all the properties for a deleted record to figure out a smart kind of report to tell users about how this impacts their features, we may want to start doing soft deletes instead of hard deletes. We can do this by adding a new date field `deletedAt` to the collection, and updating all existing queries to filter by null `deletedAt`. It may also make sense to insert a revision and in the Slack handler, adjust messaging accordingly.